src/persistQueryClient/index.ts

```diff
@@ -84,8 +90,53 @@ export async function persistQueryClient({
       )
       persister.removeClient()
     }
+  }
+}
+
+/**
+ * Persists data from the QueryCache
+ *  - data dehydrated using dehydrateOptions
+ *  - data is persisted using persister.persistClient
+ */
+export async function persistQueryClientSave({
+  queryClient,
+  persister,
+  buster = '',
+  dehydrateOptions,
+}: PersistedQueryClientSaveOptions) {
+  if (typeof window !== 'undefined') {
+    const persistClient: PersistedClient = {
+      buster,
+      timestamp: Date.now(),
+      clientState: dehydrate(queryClient, dehydrateOptions),
+    }
+
+    await persister.persistClient(persistClient)
+  }
+}
+
+/**
+ * Subscribe to QueryCache updates (for persisting)
+ * @returns an unsubscribe function (to discontinue monitoring)
+ */
+export function persistQueryClientSubscribe(
+  props: PersistedQueryClientSaveOptions
+) {
+  return props.queryClient.getQueryCache().subscribe(() => {
+    persistQueryClientSave(props)
+  })
+}
+
+/**
+ * Restores persisted data to QueryCache and persists further changes.
+ * (Retained for backwards compatibility)
+ */
+export async function persistQueryClient(props: PersistQueryClientOptions) {
+  if (typeof window !== 'undefined') {
+    // Attempt restore
+    persistQueryClientRestore(props)
 
     // Subscribe to changes in the query cache to trigger the save
-    queryClient.getQueryCache().subscribe(saveClient)
+    persistQueryClientSubscribe(props)
```

Comment:
should we not `return persistQueryClientSubscribe(props)` here in order to fix the issue of being unable to unsubscribe?

Reply:
I'm happy either way.

if a user wants to control the subscribe/unsubscribe, my assumption was that they would call the new `persistQueryClientSubscribe` directly.  The way I submitted, `persistQueryClient` works **exactly** as before.  But I perhaps some would find it nice to also add the feature here.  Let me know if you want me to add this to the PR.

Reply:
yes lets please add / fix that. We're towards a new major version anyways :)

Reply:
done: https://github.com/tannerlinsley/react-query/pull/3141/commits/c4651b1851745f9a03f92deb392d1c2778de33f6